source/module_elecstate/module_dm/density_matrix_io.cpp

```diff
@@ -0,0 +1,446 @@
+#include "density_matrix.h"
+
+#include "module_parameter/parameter.h"
+#include "module_base/libm/libm.h"
+#include "module_base/memory.h"
+#include "module_base/timer.h"
+#include "module_base/tool_title.h"
+#include "module_cell/klist.h"
+
+namespace elecstate
+{
+
+// initialize density matrix DMR from UnitCell (mainly used in UnitTest)
+template <typename TK, typename TR>
+void DensityMatrix<TK, TR>::init_DMR(Grid_Driver* GridD_in, const UnitCell* ucell)
+{
+    ModuleBase::TITLE("DensityMatrix", "init_DMR");
+    // ensure _DMR is empty
+    for (auto& it: this->_DMR)
+    {
+        delete it;
+    }
+    this->_DMR.clear();
+    // construct a new DMR
+    hamilt::HContainer<TR>* tmp_DMR;
+    tmp_DMR = new hamilt::HContainer<TR>(this->_paraV);
+    // set up a HContainer
+    for (int iat1 = 0; iat1 < ucell->nat; iat1++)
+    {
+        auto tau1 = ucell->get_tau(iat1);
+        int T1, I1;
+        ucell->iat2iait(iat1, &I1, &T1);
+        AdjacentAtomInfo adjs;
+        GridD_in->Find_atom(*ucell, tau1, T1, I1, &adjs);
+        // std::cout << "adjs.adj_num: " <<adjs.adj_num << std::endl;
+        for (int ad = 0; ad < adjs.adj_num + 1; ++ad)
+        {
+            const int T2 = adjs.ntype[ad];
+            const int I2 = adjs.natom[ad];
+            int iat2 = ucell->itia2iat(T2, I2);
+            if (this->_paraV->get_row_size(iat1) <= 0 || this->_paraV->get_col_size(iat2) <= 0)
+            {
+                continue;
+            }
+            ModuleBase::Vector3<int>& R_index = adjs.box[ad];
+            // std::cout << "R_index: " << R_index.x << " " << R_index.y << " " << R_index.z << std::endl;
+            hamilt::AtomPair<TR> tmp_ap(iat1, iat2, R_index, this->_paraV);
+            tmp_DMR->insert_pair(tmp_ap);
+        }
+    }
+    // allocate the memory of BaseMatrix in SR, and set the new values to zero
+    if (std::is_same<TK, double>::value)
+    {
+        tmp_DMR->fix_gamma();
+    }
+    tmp_DMR->allocate(nullptr, true);
+    this->_DMR.push_back(tmp_DMR);
+    // add another DMR if nspin==2
+    if (this->_nspin == 2)
+    {
+        hamilt::HContainer<TR>* tmp_DMR1;
+        tmp_DMR1 = new hamilt::HContainer<TR>(*tmp_DMR);
```

Comment:
where is 'delete'?

Reply:
This code come from task "separate file density_matrix.cpp to avoid larger than 500 lines files", but not related with the main task "Fix" of this PR, I'm glad to discuss these code elsewhere than in this PR, the remaining issues from refactoring of "density_matrix.cpp" should not block this Fix PR.